src/datanode/src/sql/delete.rs

```diff
@@ -0,0 +1,146 @@
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::ops::Deref;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use common_query::Output;
+use common_telemetry::warn;
+use datatypes::prelude::VectorRef;
+use datatypes::vectors::{BooleanVector, Float64Vector, StringVector, TimestampMillisecondVector};
+use session::context::QueryContextRef;
+use snafu::ResultExt;
+use sql::ast::{BinaryOperator, Expr, Value};
+use sql::statements::delete::Delete;
+use table::engine::TableReference;
+use table::requests::DeleteRequest;
+
+use crate::error::{DeleteSnafu, InvalidSqlSnafu, Result};
+use crate::instance::sql::table_idents_to_full_name;
+use crate::sql::{SqlHandler, SqlRequest};
+
+impl SqlHandler {
+    pub(crate) async fn delete(&self, req: DeleteRequest) -> Result<Output> {
+        let table_ref = TableReference {
+            catalog: &req.catalog_name.to_string(),
+            schema: &req.schema_name.to_string(),
+            table: &req.table_name.to_string(),
+        };
+
+        let table = self.get_table(&table_ref)?;
+
+        let affected_rows = table.delete(req).await.with_context(|_| DeleteSnafu {
+            table_name: table_ref.to_string(),
+        })?;
+
+        Ok(Output::AffectedRows(affected_rows))
+    }
+
+    pub(crate) fn delete_to_request(
+        &self,
+        stmt: Delete,
+        query_ctx: QueryContextRef,
+    ) -> Result<SqlRequest> {
+        let (catalog_name, schema_name, table_name) =
+            table_idents_to_full_name(stmt.table_name(), query_ctx.clone())?;
+        let key_column_values = parser_selection(stmt.selection())?;
+        Ok(SqlRequest::Delete(DeleteRequest {
+            key_column_values,
+            catalog_name,
+            schema_name,
+            table_name,
+        }))
+    }
+}
+
+/// parser selection, currently supported format is `tagkey1 = 'tagvalue1' and tagkey2 = 'tagvalue2'`.
+/// (only uses =, and in the where clause and provides all columns needed by the key.)
+fn parser_selection(selection: &Option<Expr>) -> Result<HashMap<String, VectorRef>> {
+    let mut key_column_values = HashMap::new();
+    match selection {
+        Some(expr) => {
+            parser_expr(&expr, &mut key_column_values)?;
+        }
+        _ => {}
+    }
+    return Ok(key_column_values);
+}
+
+fn parser_expr(expr: &Expr, key_column_values: &mut HashMap<String, VectorRef>) -> Result<()> {
+    // match BinaryOp
+    if let Expr::BinaryOp { left, op, right } = expr {
+        // match And operator
+        if let BinaryOperator::And = op {
+            if let Expr::BinaryOp { .. } = left.deref() {
+                if let Expr::BinaryOp { .. } = right.deref() {
+                    parser_expr(left.deref(), key_column_values)?;
+                    parser_expr(right.deref(), key_column_values)?;
+                    return Ok(());
+                }
+            }
+        }
+        // match eq operator
+        else if let BinaryOperator::Eq = op {
```

Comment:
I personally prefer `match` over chained `if-let` for this case.

Reply:
In this function block, there may be many situations that need to be ignored. Perhaps using let match may be easier to understand?

Reply:
```rust
    match expr {
        Expr::BinaryOp { left, op, right } => {
            match op {
                BinaryOperator::And => {
                    match left.deref() {
                        Expr::BinaryOp { .. } => {
                            match right.deref() {
                                Expr::BinaryOp { .. } => {
                                    parser_expr(left.deref(), key_column_values)?;
                                    parser_expr(right.deref(), key_column_values)?;
                                    return Ok(());
                                }
                                _ => {}
                            }
                        }
                        _ => {}
                    }
                }
                BinaryOperator::Eq => {
                    match left.deref() {
                        Expr::Identifier(column_name) => {
                            match right.deref() {
                                Expr::Value(value) => {
                                    key_column_values.insert(
                                        column_name.to_string(),
                                        value_to_vector(&column_name.to_string(), &value)?,
                                    );
                                    return Ok(());
                                }
                                Expr::Identifier(value) => {
                                    key_column_values.insert(
                                        column_name.to_string(),
                                        Arc::new(StringVector::from(vec![value.to_string()])),
                                    );
                                    return Ok(());
                                }
                                _ => {}
                            }
                        }
                        _ => {}
                    }
                }
                _ => {}
            }
        }
        _ => {}
    }
```
If match is used here, it looks very complicated.

Reply:
How about this:
```rust
    match expr {
        if let Expr::BinaryOp { left, op, right } = expr {
            match (left, op, right) {
                (Expr::BinaryOp { .. }, BinaryOperator::And, Expr::BinaryOp { .. }) => {
                    parser_expr(left.deref(), key_column_values)?;
                    parser_expr(right.deref(), key_column_values)?;
                     Ok(());
                },
                (Expr::Identifier(column_name), BinaryOperator::Eq, Expr::Value(value)) => {
                    key_column_values.insert(
                        column_name.to_string(),
                        value_to_vector(&column_name.to_string(), &value)?,
                    );
                     Ok(());
                },
            }
        }
    }
```

Reply:
wooow, thanks a lot!